apps/microtvm/zephyr/template_project/src/aot_standalone_demo/platform.c

```diff
@@ -16,19 +16,64 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-#include "zephyr_uart.h"
 
+#include "tvm/platform.h"
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
 #include <zephyr/drivers/uart.h>
+#include <zephyr/sys/reboot.h>
 #include <zephyr/sys/ring_buffer.h>
 
 #include "crt_config.h"
+#include "dlpack/dlpack.h"
+#include "tvm/runtime/crt/error_codes.h"
+#include "tvmgen_default.h"
 
 static const struct device* g_microtvm_uart;
 #define RING_BUF_SIZE_BYTES (TVM_CRT_MAX_PACKET_SIZE_BYTES + 100)
 
 // Ring buffer used to store data read from the UART on rx interrupt.
 RING_BUF_DECLARE(uart_rx_rbuf, RING_BUF_SIZE_BYTES);
 
+void TVMLogf(const char* msg, ...) {
+  char buffer[256];
+  int size;
+  va_list args;
+  va_start(args, msg);
+  size = vsprintf(buffer, msg, args);
+  va_end(args);
+  TVMPlatformWriteSerial(buffer, (uint32_t)size);
+}
+
+// Called by TVM when a message needs to be formatted.
+__attribute__((weak)) size_t TVMPlatformFormatMessage(char* out_buf, size_t out_buf_size_bytes,
+                                                      const char* fmt, va_list args) {
+  return vsnprintk(out_buf, out_buf_size_bytes, fmt, args);
+}
+
+// Called by TVM when an internal invariant is violated, and execution cannot continue.
+__attribute__((weak)) void TVMPlatformAbort(tvm_crt_error_t error) {
+  TVMLogf("TVMPlatformAbort: %08x\n", error);
+  sys_reboot(SYS_REBOOT_COLD);
+  for (;;)
+    ;
```

Comment:
Does `sys_reboot` ever return? I'm not sure we need this infinite loop.

Reply:
I'm not sure how does it work, I rather make this change when we test it on a physical board

Reply:
i think it helps with codegen because the compiler understands that `for (;;) ;` never returns, but not necessarily that sys_reboot doesn't. but honestly i don't know whether it's needed (there's an attribute noreturn that can be added to sys_reboot that does the same, but i believe i've seen errors from that not being recognized before).